domain/modelagent/service/service.go

```diff
@@ -24,43 +52,247 @@ type Service struct {
+// - [github.com/juju/juju/domain/model/errors.NotFound] - When the model the
+// unit belongs to no longer exists.
+func (s *ModelService) GetUnitTargetAgentVersion(
+	ctx context.Context,
+	unitName string,
+) (version.Number, error) {
+	err := s.st.CheckUnitExists(ctx, unitName)
+	if errors.Is(err, applicationerrors.UnitNotFound) {
+		return version.Zero, errors.Errorf(
+			"unit %q does not exist", unitName,
+		).Add(applicationerrors.UnitNotFound)
+	} else if err != nil {
+		return version.Zero, errors.Errorf(
+			"checking if unit %q exists when getting target agent version: %w",
+			unitName, err,
+		)
+	}
+
+	modelUUID, err := s.st.GetModelUUID(ctx)
+	if err != nil {
+		return version.Zero, errors.Errorf(
+			"getting unit %q model uuid: %w", unitName, err,
+		)
+	}
+
+	return s.Service.st.GetModelTargetAgentVersion(ctx, modelUUID)
+}
+
+// WatchApplicationTargetAgentVersion is responsible for watching the target
+// agent version for an application and reporting when there has been a change
+// via a [watcher.NotifyWatcher]. The following errors can be expected:
+// - [applicationerrors.ApplicationNotFound] - When the application does not
+// exist.
+// - [modelerrors.NotFound] - When the model of the machine no longer exists.
+func (s *ModelService) WatchApplicationTargetAgentVersion(
+	ctx context.Context,
+	applicationName string,
+) (watcher.NotifyWatcher, error) {
+	err := s.st.CheckApplicationExists(ctx, applicationName)
+	if errors.Is(err, applicationerrors.ApplicationNotFound) {
+		return nil, errors.Errorf(
+			"application %q does not exist", applicationName,
+		).Add(applicationerrors.ApplicationNotFound)
+	} else if err != nil {
+		return nil, errors.Errorf(
+			"checking if application %q exists when watching target agent version: %w",
+			applicationName, err,
+		)
+	}
+
+	w, err := s.WatchModelTargetAgentVersion(ctx)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"getting watcher for application %q model target agent version: %w",
+			applicationName, err,
+		)
+	}
+	return w, nil
+}
+
+// WatchMachineTargetAgentVersion is responsible for watching the target agent
+// version for machine and reporting when there has been a change via a
+// [watcher.NotifyWatcher]. The following errors can be expected:
+// - [machineerrors.NotFound] - When no machine exists for the provided name.
+// - [modelerrors.NotFound] - When the model of the machine no longer exists.
+func (s *ModelService) WatchMachineTargetAgentVersion(
+	ctx context.Context,
+	machineName machine.Name,
+) (watcher.NotifyWatcher, error) {
+	err := s.st.CheckMachineExists(ctx, machineName)
+	if errors.Is(err, machineerrors.MachineNotFound) {
+		return nil, errors.Errorf(
+			"machine %q does not exist", machineName,
+		).Add(machineerrors.MachineNotFound)
+	} else if err != nil {
+		return nil, errors.Errorf(
+			"checking if machine %q exists when watching target agent version: %w",
+			machineName, err,
+		)
+	}
+
+	w, err := s.WatchModelTargetAgentVersion(ctx)
```

Comment:
Is this correct? When you watch a machine, are you watching the model in terms of the agent version?

----

Edit: I guess it is.

Reply:
Yeah it is. Same as what we had before just more explicit about what is happening. We hide implementation from the API server now.